### PR TITLE
Display loading progress bar for the Segments list table

### PR DESCRIPTION
**Pull request by zackcl**
This PR displays the loading progress bar for the Segments list table while fetching the segments from the backend. Additionally, the Segments list is now only visible after the loading/sorting is complete (previously, it initially displayed the unsorted list for half a second).

### CODE REVIEW

File: frontend/projects/upgrade/src/app/core/segments/segments.service.ts [hunk 1]
```diff
@@ -125,6 +125,7 @@ export class SegmentsService {
   }
 
   fetchSegments(fromStarting?: boolean) {
+    this.store$.dispatch(SegmentsActions.actionSetIsLoadingSegments({ isLoadingSegments: true }));
```
Comment (VivekFitkariwala) [inline frontend/projects/upgrade/src/app/core/segments/segments.service.ts:128]: This is not required. The handler for actionFetchSegments is not implmented.

**Review by VivekFitkariwala — CHANGES_REQUESTED** [review]

File: frontend/projects/upgrade/src/app/core/segments/segments.service.ts (hunk 1 above)
Comment (zackcl) [reply frontend/projects/upgrade/src/app/core/segments/segments.service.ts:128]: What do you mean? I believe the handlers for `actionFetchSegments` are implemented in the effect:

```
fetchSegments$ = createEffect(() =>
    this.actions$.pipe(
      ofType(SegmentsActions.actionFetchSegments),
      withLatestFrom(this.store$.pipe(select(selectAllSegments))),
      switchMap(() =>
        this.segmentsDataService.fetchSegments().pipe(
          map((data: any) =>
            SegmentsActions.actionFetchSegmentsSuccess({
              segments: data.segmentsData,
              experimentSegmentInclusion: data.experimentSegmentInclusionData,
              experimentSegmentExclusion: data.experimentSegmentExclusionData,
            })
          ),
          catchError(() => [SegmentsActions.actionFetchSegmentsFailure()])
        )
      )
    )
  );
```

And also in the reducer:

```
const reducer = createReducer(
  initialState,
  on(SegmentsActions.actionUpsertSegment, SegmentsActions.actionGetSegmentById, (state) => ({
    ...state,
    isLoadingSegments: true,
  })),
  on(
    SegmentsActions.actionFetchSegmentsSuccess,
    (state, { segments, experimentSegmentExclusion, experimentSegmentInclusion }) => {
      const newState = {
        ...state,
        segments,
        allExperimentSegmentsInclusion: experimentSegmentInclusion,
        allExperimentSegmentsExclusion: experimentSegmentExclusion,
      };
      return adapter.upsertMany(segments, { ...newState, isLoadingSegments: false });
    }
  ),
  on(
    SegmentsActions.actionFetchSegmentsFailure,
    SegmentsActions.actionUpsertSegmentFailure,
    SegmentsActions.actionGetSegmentByIdFailure,
    (state) => ({ ...state, isLoadingSegments: false })
  ),
  on(SegmentsActions.actionUpsertSegmentSuccess, (state, { segment }) =>
    adapter.upsertOne(segment, { ...state, isLoadingSegments: false })
  ),
  on(SegmentsActions.actionGetSegmentByIdSuccess, (state, { segment }) =>
    adapter.upsertOne(segment, { ...state, isLoadingSegments: false })
  ),
  on(SegmentsActions.actionSetSearchKey, (state, { searchKey }) => ({ ...state, searchKey })),
  on(SegmentsActions.actionSetSearchString, (state, { searchString }) => ({ ...state, searchString })),
  on(SegmentsActions.actionSetSortKey, (state, { sortKey }) => ({ ...state, sortKey })),
  on(SegmentsActions.actionSetSortingType, (state, { sortingType }) => ({ ...state, sortAs: sortingType })),
  on(SegmentsActions.actionDeleteSegmentSuccess, (state, { segment }) => adapter.removeOne(segment.id, state)),
  on(SegmentsActions.actionSetIsLoadingSegments, (state, { isLoadingSegments }) => ({ ...state, isLoadingSegments }))
);
```

If I remove the `this.store$.dispatch(SegmentsActions.actionSetIsLoadingSegments({ isLoadingSegments: true }));` the loading progress bar doesn't appear on the Segments list table. Therefore, I believe this line is required. Let me know if you have different suggestions.

Comment (VivekFitkariwala) [reply frontend/projects/upgrade/src/app/core/segments/segments.service.ts:128]: The reducer for `actionFetchSegments` is not there.

It should be there in the `segments.reducer.ts`
`on(SegmentsActions.actionUpsertSegment, SegmentsActions.actionGetSegmentById, (state) => ({`

You should not call `actionSetIsLoadingSegments` explicitly.

Comment (zackcl) [reply frontend/projects/upgrade/src/app/core/segments/segments.service.ts]: Thanks, I applied the change you suggested. Also, the sorting is updated to be case-insensitive (just like how the experiments and flags list tables work).

**Comment by zackcl** [comment]
@VivekFitkariwala Please review and approve this PR (not sure why I can't see the "re-request a review" button).

**Review by VivekFitkariwala — APPROVED** [review]

**Pull request merged**